src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogHttpClient.java

```diff
@@ -76,28 +90,51 @@ public boolean sendEvent(JSONObject payload) {
 
     @Override
     public void incrementCounter(String name, String hostname, JSONArray tags) {
-        ConcurrentMap<CounterMetric, Integer> counters = ConcurrentMetricCounters.Counters.get();
-        CounterMetric counterMetric = new CounterMetric(tags, name, hostname);
-        Integer previousValue = counters.putIfAbsent(counterMetric, 1);
-        if (previousValue != null){
-            boolean ok = counters.replace(counterMetric, previousValue, previousValue + 1);
-            while(!ok) {
-                previousValue = counters.get(counterMetric);
-                ok = counters.replace(counterMetric, previousValue, previousValue + 1);
+        ConcurrentMetricCounters.lock.lock();
+        try {
+            ConcurrentMetricCounters countersInstance = ConcurrentMetricCounters.getInstance();
+            ConcurrentMap<CounterMetric, Integer> counters = countersInstance.get();
+            CounterMetric counterMetric = new CounterMetric(tags, name, hostname);
+            Integer previousValue = counters.putIfAbsent(counterMetric, 1);
+            if (previousValue != null){
+                boolean ok = counters.replace(counterMetric, previousValue, previousValue + 1);
+                // NOTE:
+                // This while loop below should never be called since we are using a lock when flushing and
+                // incrementing counters.
+                while(!ok) {
+                    logger.warning("Couldn't increment counter " + name + " with value " + (previousValue + 1) +
+                            " previousValue = " + previousValue);
+                    previousValue = counters.get(counterMetric);
+                    ok = counters.replace(counterMetric, previousValue, previousValue + 1);
+                }
             }
+            previousValue = previousValue == null ? 0 : previousValue;
+            logger.fine("Counter " + name + " updated from previousValue " + previousValue + " to "
+                    + (previousValue + 1));
+        } finally {
+            ConcurrentMetricCounters.lock.unlock();
         }
     }
 
     @Override
     public void flushCounters() {
-        ConcurrentMap<CounterMetric, Integer> counters = ConcurrentMetricCounters.Counters.get();
-        ConcurrentMetricCounters.Counters.set(new ConcurrentHashMap<CounterMetric, Integer>());
-
+        ConcurrentMetricCounters.lock.lock();
+        ConcurrentMap<CounterMetric, Integer> counters;
+        try {
+            ConcurrentMetricCounters countersInstance = ConcurrentMetricCounters.getInstance();
+            counters = countersInstance.get();
+            countersInstance.reset();
```

Comment:
Same here, if you had a synchronized method like `getAndReset`, you could get rid of the manual lock

Reply:
see my comment above. I like the `getAndReset` method idea though :)

Reply:
Actually i won't do the `getAndReset` method. I initially thought I would have one in order to group the following logic: 
```
ConcurrentMetricCounters countersInstance = ConcurrentMetricCounters.getInstance();
counters = countersInstance.get();
```
Maybe i will create a Utilities method in another PR